Fixed `uuid!("")` macro diagnostics when using `urn:uuid` prefix.

QnnOkabayashi opened the pull request:
**I'm submitting a** bug fix

# Description
Fixed an bug where the `uuid!("")` macro would point to the wrong characters when using the `urn:uuid:` prefix or wrapping with the curly braces (`"{...}"`).

# Related Issue(s)
none

KodrAus commented:
Nice catch!

Looks like we’ve got a couple test failures here. I think the `start` variable is expected to be fixed, so we might need to track the index to report errors at from a different starting point?

We also just need to make sure the parser will never panic. Maybe that should go in a comment at the top of the function?

QnnOkabayashi commented:
> Nice catch!
> 
> Looks like we’ve got a couple test failures here. I think the `start` variable is expected to be fixed, so we might need to track the index to report errors at from a different starting point?
> 
Fixed the indexing error that was causing usize overflow
> We also just need to make sure the parser will never panic. Maybe that should go in a comment at the top of the function?

I also found out that it panicked when a wide char was present, so I had to add some support in the macro that fixed this.

Also, I added some util functions for creating `Error`s to clean up the parsing algorithm a bit.

KodrAus commented on shared/parser.rs:
```diff
@@ -127,21 +124,17 @@ pub fn parse_str(mut input: &str) -> Result<[u8; 16], Error> {
-                    }
-                    .into());
+                    // let found = input[i_char..].chars().next().unwrap();
+                    // let found = char::from(chr);
```
Did we mean to keep these comments here?

QnnOkabayashi replied:
Nope that was an accident

KodrAus commented on tests/ui/compile_fail/invalid_parse.rs:
```diff
@@ -25,5 +25,18 @@ const _: Uuid = uuid!("F9168C5E-CEB2-4faa-BBF-329BF39FA1E4");
+const _: Uuid = uuid!("67e550-4105b1426f9247bb680e5fe0c");
+
+const _: Uuid = uuid!("504410岡林aab1426f9247bb680e5fe0c8");
+const _: Uuid = uuid!("504410😎👍aab1426f9247bb680e5fe0c8");
```
😎 👍

KodrAus commented:
That is much easier to follow. Do we have any test cases for the `parse_str` function itself that includes wide characters?

QnnOkabayashi commented:
> That is much easier to follow. Do we have any test cases for the `parse_str` function itself that includes wide characters?

Not yet. Also how optimized is the parsing algorithm? The more I look at it, the more it bothers me how clunky it feels, and I'd like to refactor it if you're not opposed.

KodrAus commented:
> I'd like to refactor it if you're not opposed.

I think it's had one or two rounds of optimization applied in its lifetime, but nothing like a complete rewrite as far as I remember. If you'd like to spend some time on it that sounds fantastic! I don't mind if you just want to keep pushing changes to this PR here, or if we merge and start another.
